bin/cheritest/cheritest_cheriabi.c

```diff
@@ -251,3 +251,133 @@ test_cheriabi_malloc_zero_size(const struct cheri_test *ctp __unused)
 		    "non-zero length (%zu)", cheri_getlength(cap));
 	cheritest_success();
 }
+
+struct adjacent_mappings {
+	char *first;
+	char *middle;
+	char *last;
+	size_t maplen;
+};
+
+/*
+ * Create three adjacent memory mappings that be used to check that the memory
+ * mapping system calls reject out-of-bounds capabilities that have the address
+ * of a valid mapping.
+ */
+static void
+create_adjacent_mappings(struct adjacent_mappings *mappings)
+{
+	char *requested_addr;
+	size_t len;
+
+	len = getpagesize() * 2;
+	mappings->first = CHERITEST_CHECK_SYSCALL(
+	    mmap(0, len, PROT_READ | PROT_WRITE, MAP_ANON, -1, 0));
+	CHERITEST_VERIFY(cheri_gettag(mappings->first));
+	/* Try to create a mapping immediately following the latest one. */
+	requested_addr = cheri_cleartag(mappings->first) + len;
```

Comment:
We may be about to break this in the purecap reservations branch by disallowing non-NULL-derived arguments (in part due to my worry that people creating adjacent mappings probably mean to assume the original pointer can access the whole thing). 

cc: @qwattash

Reply:
Don't we need MAP_FIXED for bss in RTLD?

Reply:
RTLD walks the phdrs before doing the big mapping, so its bss mapping code still has a capability for those pages.

Reply:
As things stand, you will have to modify code that tries to make adjacent mappings, the requirement that `addr` be NULL-derived simply changes what modifications you need to make (e.g. you need to do `(void *)(uintcap_t)cheri_getaddress(ptr)` instead of `cheri__cleartag(ptr)`.

My main reason for wanting to reject non-NULL-derived capabilities is that you risk weird and inconsistent results in the face of revocation.

(As a side note, I think I'd like an eventual cmmap have a notion of "allocator" that can be associated with properties like high-address, traditional-heap, low-address, etc both for this sort of use case and to potentially provide hints as the the source of addresses in the debugger. This will be especially useful in a co-process world where I expect addresses to be intermixed aggressively, particularly as the address space fragments with long-running processes.)

Reply:
I completely misunderstood your comment as "the reservations changes will remove MAP_FIXED" rather than "MAP_FIXED hints need to be NULL derived". I agree that the latter makes sense and I'll update the change.